Tls Passthrough support

zijun726911 opened the pull request:
Initial version cherry picked from: https://github.com/liwenwu-amazon/aws-application-networking-k8-publics/tree/tls-route-support-mar20
User experiences about TLSRoute, TargetGroupPolicy of this PR are same with the `liwenwu-amazon:tls-route-support-mar20 branch`. 

`main branch` to  `liwenwu-amazon:tls-route-support-mar20 branch` diff: https://github.com/zijun726911/aws-application-networking-k8s/pull/3

## TLS Passthrough Behaviors Summary
- Users can create `Kind:TLSRoute` k8s resource, the controller will under the hood create VPC Lattice Service with TLS_PASSTHROUGH protocol listener and TCP type TargetGroup.
- For cross cluster use case, users can create `Kind:ServiceExport`  k8s resource and create the `Kind:TargetGroupPolicy` resource to specify the TargetGroup protocol to TCP. The controller will under the hood create VPC Lattice TCP TargetGroup
- By default, controller created TCP TargetGroups has healthcheck==disabled
- Same as old TargetGroupPolicy use cases, users can create the `Kind:TargetGroupPolicy`  to override the healthcheck config for either Service (backendref by TLSRoute) or ServiceExport
- For the TargetGroupPolicy with protocol==TCP && protocolVersion!=nil, it will be rejected.
- The controller created TCP Target Groups have tag: `application-networking.k8s.aws/ProtocolVersion:""`

## Changes

### The controller logic changes:
- Make the e2e code path (route_controller/serviceexport_controller --> model_build_xxxxx -->  xxx_synthesizer --> target_group_manager) support the TLSRoute, and the ServiceExport with `protocol:TCP` TargetGroupPolicy.
- In the `rule_synthesizer.go`, skips to create rules for TLSRoute since Lattice TLS_PASSTHROUGH listener can only have one defaultAction and without any addition rules. In the `listener_synthesizer.go`, fill the lattice listener's defaultAction to point to the referred TCP TargetGroups.
- To support Tls passthrough, both `rule_synthesizer` and `listener_synthesizer` need the `ResolveRuleTgIds()` function (originally resided in the rule_synthesizer.go), so I move the `ResolveRuleTgIds()`  and `findSvcExportTG()` to the `target_group_manager.go` to avoid copy paste same code, and the `rule_synthesizer` and `listener_synthesizer` can both use the same `s.tgManager.ResolveRuleTgIds(ctx, rule, r.stack)`
- **Encapsulate TLS_PATHTHROUGH listener related logic in the listenerSynthesizer.getLatticeListenerDefaultAction()**
- Move the unit test `Test_ResolveRuleTgIds()` to the target_group_manager_test.go
- Set healthcheck to disabled for TCP targetGroup by default (for both target_group_manager `Create()` and `Update()` methods). If the user specify the healthcheck setting by TargetGroupPolicy, the default healthcheck setting will be overridden. (Disable healthcheck by default for TCP TargetGroup is also the default VPC Lattice API default behavior) https://docs.aws.amazon.com/vpc-lattice/latest/ug/target-group-health-checks.html
- Since VPC Lattice TCP TargetGroup don't have ProtocolVersion,  **for the TCP TargetGroup, set the tag `application-networking.k8s.aws/ProtocolVersion:""`**.  In the `(t *TargetGroupSpec) Validate()` remove the `ProtocolVersion` from `requiredFields`.
- Improvement for targetgrouppolicy with `targetRef ServiceExport`, now, the targetgrouppolicy_controller `Watch()` the ServiceExport resource change

### e2e test changes
- Add new automation e2e tests to test TLSRoute and ServiceExport with `protocol:TCP` TargetGroupPolicy

### CRDs && Image build && helm chart related changes
- Added tlsroutes  create, delete, get, list, patch, update, watch permissions for the controller serviceAccount AuthZ permissions
- Added gateway.networking.k8s.io_tlsroutes.yaml, included the tlsroutes CRDs in the helm chart and deploy.yaml 


## Testing

- Following tls-passthrough.md steps to set up the single cluster e2e connectivity (TLSRoute backendref Service) and cross-cluster e2e connectivity (TLSRoute backendref ServiceImport + ServiceExport + TargetGroupPolicy) both work as expected.
- 2 new automation e2e tests can pass (one test TLSRoute and the other one test the TLSRoute+ ServiceImport+ ServiceExport + TargetGroupPolicy).
- Ran the whole e2e test suite, all test can pass (skiped the RAM share test)

```
[SynchronizedAfterSuite]
/Volumes/workspace/aws-application-networking-k8s/test/suites/integration/suite_test.go:72
{“level”:“info”,“ts”:“2024-06-04T01:32:32.262-0700",“caller”:“test/framework.go:264",“msg”:“Deleting objects: *v1.Gateway/test-gateway, *v1.Pod/grpc-runner”}
{“level”:“info”,“ts”:“2024-06-04T01:32:32.312-0700",“caller”:“test/framework.go:282",“msg”:“Waiting for NotFound, objects: *v1.Gateway/test-gateway, *v1.Pod/grpc-runner”}
[SynchronizedAfterSuite] PASSED [31.091 seconds]
------------------------------
Ran 66 of 69 Specs in 2537.133 seconds
SUCCESS! -- 66 Passed | 0 Failed | 0 Pending | 3 Skipped
```


### Changes for the new revision: https://github.com/aws/aws-application-networking-k8s/pull/643#discussion_r1629937229


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

zijun726911 commented on pkg/gateway/model_build_listener.go:
```diff
@@ -28,33 +30,29 @@ func (t *latticeServiceModelBuildTask) extractListenerInfo(
 	if err != nil {
```
Just to make the `extractListenerInfo()` more easy to follow (by reversing the `parentRef.SectionName != nil` condition and removing the `found := false` flag and returning eariler ).


 and override protocol string to "TLS_PASSTHROUGH" when create the lattice listener for k8s Gateway with `section.TLS.Mode==Passthrough`

liwenwu-amazon commented on .github/workflows/publish-doc.yaml:
```diff
@@ -26,7 +26,7 @@ jobs:
       - name: Build
         run: |
-          mike deploy 1.0.5 latest --update-aliases --push
+          mike deploy v1.0.6-rc1 latest --update-aliases --push
```
These changes seems not related to TLS passthrough feature.  Can you separate them into different CR? thanks

zijun726911 replied:
This one is a little bit special, need to bump up the doc version here. otherwise the the doc version 1.0.5 will include the tls-passthrough.md, but the the controller v1.0.5 actually don't support the TLSRoute.

https://www.gateway-api-controller.eks.aws.dev/1.0.5/


What I am think is we can improve the whole Github Action (CI pipeline) publish-doc.yaml (in the future),  for example, main branch code change trigger the `mike deploy dev latest`, and branch `release-xxxxx` trigger the `mike deploy v1.x.x `

liwenwu-amazon reviewed (commented):
can you move non TLS-passthrough related change into a separate CR? thanks

liwenwu-amazon reviewed (approved):
LGTM

erikfuller commented on test/suites/integration/tlsroute_serviceexport_test.go:
```diff
@@ -0,0 +1,150 @@
+		Expect(*tgSummary.VpcIdentifier).To(Equal(os.Getenv("CLUSTER_VPC_ID")))
+		Expect(*tgSummary.Protocol).To(Equal("TCP"))
+		if tg.Config.HealthCheck != nil {
+			Expect(*tg.Config.HealthCheck.Enabled).To(BeFalse())
```
Since we already need a TG policy here, would it be a good idea to include an enabled health check to make sure that's being set up properly?

erikfuller commented on test/suites/integration/tlsroute_serviceexport_test.go:
```diff
@@ -0,0 +1,150 @@
+		}
+
+		targetsV1 := testFramework.GetTargets(ctx, tgSummary, deployment1)
+		Expect(*tgSummary.Port).To(BeEquivalentTo(80))
```
Wouldn't this be `443`?

zijun726911 replied:
This 80 is come from here: https://github.com/aws/aws-application-networking-k8s/blob/9cc4efb77535892ad017a480f875b5011adb3bf4/pkg/gateway/model_build_targetgroup.go#L363

The Target Group port is kind of useless, only the lattice targets' port are useful and guide vpc lattice to route traffic to that port. i.e, :


```
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: TLSRoute
metadata:
  name: rate-tls-passthrough
spec:
  hostnames:
    - tls-rate.my-test.com
  parentRefs:
    - name: my-hotel-tls-passthrough
      sectionName: tls
  rules:
    - backendRefs:
        - name: tls-rate1
          kind: Service
          port: 443. <-------- This port + k8s service ports intersection define to the lattice target port
          weight: 10
```

erikfuller replied:
Should we just drop this `Expect()` then since we don't really care?

erikfuller commented on test/suites/integration/tlsroute_serviceexport_test.go:
```diff
@@ -0,0 +1,150 @@
+		pod := pods[0]
+
+		Eventually(func(g Gomega) {
+			cmd := fmt.Sprintf("curl -k https://tls.test.com:444 --resolve tls.test.com:444:%s", dnsIP[0])
```
Should we have an explicit check for port `444` on the listener?

erikfuller commented on test/suites/integration/tlsroute_test.go:
```diff
@@ -0,0 +1,106 @@
+			Expect(*tg.Config.HealthCheck.Enabled).To(BeFalse())
+		}
+		targetsV1 := testFramework.GetTargets(ctx, tgSummary, deployment1)
+		Expect(*tgSummary.Port).To(BeEquivalentTo(80))
```
443?

erikfuller replied:
Same here, should we drop the `Expect()`?

zijun726911 commented on pkg/controllers/route_controller.go:
```diff
@@ -145,10 +148,9 @@ func RegisterAllRouteControllers(
 }
 
-//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=grpcroutes;httproutes,verbs=get;list;watch;create;update;patch;delete
-//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=grpcroutes/status;httproutes/status,verbs=get;update;patch
-//+kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=grpcroutes/finalizers;httproutes/finalizers,verbs=update
-
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=tlsroutes;grpcroutes;httproutes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=tlsroutes/status;grpcroutes/status;httproutes/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=gateway.networking.k8s.io,resources=tlsroutes/finalizers;grpcroutes/finalizers;httproutes/finalizers,verbs=update
```
The  +kubebuilder:rbac instructs the `controller-gen rbac` command to generate the `ClusterRole`, however our Makefile don't run the  `controller-gen rbac`, I think they are useless, will remove them

https://book.kubebuilder.io/reference/controller-gen

erikfuller replied:
Yeah that's what I thought. Just wanted to double check. Thanks for checking.

Mengli-Shu commented on pkg/deploy/lattice/target_group_manager.go:
```diff
@@ -56,15 +60,20 @@ func (s *defaultTargetGroupManager) Upsert(
 		ipAddressType = &modelTg.Spec.IpAddressType
 	}
+	if modelTg.Spec.ProtocolVersion == "" {
+		protocolVersion = nil
```
It's because protocol version is not supported for tcp target groups

erikfuller replied:
Thanks for the explanation.

Mengli-Shu commented on config/crds/bases/application-networking.k8s.aws_targetgrouppolicies.yaml:
>protocolVersion:
description: The protocol version used when performing health
checks on targets. Defaults to HTTP/1.

Hc config protocol version needs to be changed too, it doesn't have a default for tcp tg, customers need to specify both health check protocol and the protocol version

zijun726911 replied:
will remove the sentence "Defaults to HTTP/1." in the next revision

zijun726911 commented on Makefile:
```diff
@@ -131,8 +131,8 @@ e2e-test: ## Run e2e tests against cluster pointed to by ~/.kube/config
 	@echo -n "Cleaning up e2e tests... "
-	@kubectl delete namespace $(e2e-test-namespace) > /dev/null 2>&1
-	@kubectl create namespace $(e2e-test-namespace) > /dev/null 2>&1
+	-@kubectl delete namespace $(e2e-test-namespace)
+	@kubectl create namespace $(e2e-test-namespace)
```
This redirect `> /dev/null 2>&1` concealed the error message. People trying to run the `make e2e-clean` in a brand new cluster are confused if they cannot see the error message "Error from server (NotFound): namespaces "e2e-test" not found" or "couldn't get current server API group list: the server has asked for the client to provide credentials". 
(That confusing once happened in our internal workshop)

For the change : `-@kubectl delete namespace $(e2e-test-namespace)` ,  `-` is used to ignore error. So even if the `kubectl delete namespace $(e2e-test-namespace)` has `exit 1` (due to namespace do not exist), it still can go head to execute the next commend : `@kubectl create namespace $(e2e-test-namespace)`

erikfuller replied:
Nice! Thanks for the explanation.

zijun726911 commented on pkg/deploy/lattice/listener_manager.go:
```diff
@@ -70,16 +72,32 @@ func (d *defaultListenerManager) Upsert(
+		d.log.Debugf("For TLS_PASSTHROUGH listener, forward to default target groups %v", latticeTGs)
+		defaultAction = vpclattice.RuleAction{
+			Forward: &vpclattice.ForwardAction{
+				TargetGroups: latticeTGs,
```
Plan to do a refactoring for this part, for example:


```
func (l *listenerSynthesizer) getLatticeServiceListenerDefaultAction(ctx context.Context, modelListenerProtocol string) (
	*vpclattice.RuleAction, error,
) {
	if modelListenerProtocol != vpclattice.ListenerProtocolTlsPassthrough {
		return &vpclattice.RuleAction{
			FixedResponse: &vpclattice.FixedResponseAction{
				StatusCode: aws.Int64(404),
			},
		}, nil
	}

	var defaultActionTgs []*model.RuleTargetGroup
	var stackRules []*model.Rule
	_ = l.stack.ListResources(&stackRules)
	if err := l.tgManager.ResolveRuleTgIds(ctx, stackRules[0], l.stack); err != nil {
		return nil, fmt.Errorf("failed to resolve rule tg ids, err = %v", err)
	}
	// Fill the default action target groups for TLS_PASSTHROUGH listener, since TLS_PASSTHROUGH listener only has the defaultAction and no extra listener rules
	defaultActionTgs = stackRules[0].Spec.Action.TargetGroups

	// Fill the defaultAction tgs for TLS_PASSTHROUGH lattice listener
	var latticeTGs []*vpclattice.WeightedTargetGroup
	for _, modelTG := range defaultActionTgs {
		latticeTG := vpclattice.WeightedTargetGroup{
			TargetGroupIdentifier: aws.String(modelTG.LatticeTgId),
			Weight:                aws.Int64(modelTG.Weight),
		}
		latticeTGs = append(latticeTGs, &latticeTG)
	}

	l.log.Debugf("For TLS_PASSTHROUGH listener, forward to default target groups %v", latticeTGs)
	return &vpclattice.RuleAction{
		Forward: &vpclattice.ForwardAction{
			TargetGroups: latticeTGs,
		},
	}, nil

}

```

and don't expose logic about TLS_PASSTHROUGH in the `func (d *defaultListenerManager) Upsert(
	ctx context.Context,
	modelListener *model.Listener,
	modelSvc *model.Service,
	defaultAction *vpclattice.RuleAction) ` 

and  `func (l *listenerSynthesizer) Synthesize(ctx context.Context) `

zijun726911 commented on pkg/deploy/lattice/listener_manager.go:
```diff
@@ -103,11 +121,16 @@ func (d *defaultListenerManager) Upsert(
 func k8sLatticeListenerName(modelListener *model.Listener) string {
+	protocol := strings.ToLower(modelListener.Spec.Protocol)
+	if modelListener.Spec.Protocol == vpclattice.ListenerProtocolTlsPassthrough {
+		protocol = "tls"
```
The lattice listener name cannot have "_" 
![image](https://github.com/aws/aws-application-networking-k8s/assets/32318664/758970e9-a427-4c7c-b50b-cb58ded269bc)


Also, in the lattice console, the protocol name shows "TLS", not TLS_PASSTHROUGH, for consistence, we add the suffix "-tls" in the lattice listener name 
![image](https://github.com/aws/aws-application-networking-k8s/assets/32318664/d6208c92-aa5f-4a51-aaf6-e7e60a167299)

zijun726911 commented on pkg/deploy/lattice/listener_manager.go:
```diff
@@ -70,16 +72,32 @@ func (d *defaultListenerManager) Upsert(
 	}
```
I think we only need to change the comments and not code here right? 

I think we still don't need to call `lattice.UpdateListener()` from any perspective. if the the TG not found (or CREATE_IN_PROGRESS), the earlier code will return the error to prevent calling the `lattice.CreateListener()`.

for TLSRoute backendref svc code path, targetGroupSynthesizer.SynthesizeCreate()  return LATTICE_RETRY first to block to call the `lattice.CreateListener()

 for serviceExport and TLSRoute backendref serviceImport code path, the tgManager.ResolveRuleTgIds() return error to block to call the `lattice.CreateListener()`

erikfuller replied:
What would happen here if I 
1. create a TLSRoute with backendRef to serviceA
2. update the TLSRoute replacing backendRef to serviceB

I think with the logic as-is we would not update the listener appropriately. We would find the listener in `findListenerByPort`, then do nothing. Instead, we should be updating the default rule to point to the target group for service B.

zijun726911 replied:
Ok, got your point, will take a look

zijun726911 replied:
I did some refactoring in the new revision, please take a look https://github.com/aws/aws-application-networking-k8s/blob/9b3f3234fc91cc74248125a55507b38d338ef623/pkg/deploy/lattice/listener_manager.go#L46-L131

erikfuller replied:
Logic-wise, what's here looks good. Resolving.

erikfuller commented:
One other question around TLSRouteRule/BackendRef from the spec: https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/tlsroute_types.go#L103

```
	// BackendRefs defines the backend(s) where matching requests should be
	// sent. If unspecified or invalid (refers to a non-existent resource or
	// a Service with no endpoints), the rule performs no forwarding; if no
	// filters are specified that would result in a response being sent, the
	// underlying implementation must actively reject request attempts to this
	// backend, by rejecting the connection or returning a 500 status code.
	// Request rejections must respect weight; if an invalid backend is
	// requested to have 80% of requests, then 80% of requests must be rejected
	// instead.
```

Just curious what our current approach does with invalid BackendRefs. I think this is probably a less vital part of the spec to honour, especially given it's still experiemental, but would be good to describe the expected behaviour here.

zijun726911 commented on pkg/deploy/lattice/listener_synthesizer_test.go:
```diff
@@ -2,21 +2,24 @@ package lattice
+	"github.com/aws/aws-application-networking-k8s/pkg/utils/gwlog"
 )
 
 func Test_SynthesizeListenerCreate(t *testing.T) {
```
Will do

zijun726911 commented:
> One other question around TLSRouteRule/BackendRef from the spec: https://github.com/kubernetes-sigs/gateway-api/blob/main/apis/v1alpha2/tlsroute_types.go#L103
> 
> ```
> 	// BackendRefs defines the backend(s) where matching requests should be
> 	// sent. If unspecified or invalid (refers to a non-existent resource or
> 	// a Service with no endpoints), the rule performs no forwarding; if no
> 	// filters are specified that would result in a response being sent, the
> 	// underlying implementation must actively reject request attempts to this
> 	// backend, by rejecting the connection or returning a 500 status code.
> 	// Request rejections must respect weight; if an invalid backend is
> 	// requested to have 80% of requests, then 80% of requests must be rejected
> 	// instead.
> ```
> 
> Just curious what our current approach does with invalid BackendRefs. I think this is probably a less vital part of the spec to honour, especially given it's still experiemental, but would be good to describe the expected behaviour here.

I will document this part. actually the current behavior is, for example for this TLSRoute: 

```
apiVersion: gateway.networking.k8s.io/v1alpha2
kind: TLSRoute
metadata:
  name: rate-tls-passthrough
spec:
  hostnames:
    - tls-rate.my-test.com
  parentRefs:
    - name: my-hotel-tls-passthrough
      sectionName: tls
  rules:
    - backendRefs:
        - name: tls-rate1
          kind: Service
          port: 443
          weight: 10
        - name: tls-rate2
          kind: ServiceImport
          port: 443
          weight: 90
``` 

If the tls-rate2 targetGroup is not ready, it will block the traffic for both tls-rate1 and tls-rate2


 ( base on the ResolveRuleTgIds() logic https://github.com/aws/aws-application-networking-k8s/blob/595fff8ab63ef901b81b7ce0d63a2461be4d1865/pkg/deploy/lattice/target_group_manager.go#L459 )

zijun726911 commented on pkg/model/core/tlsroute.go:
```diff
@@ -0,0 +1,260 @@
+	s.Parents()[0].ControllerName = controllerName
+}
+
+func (s *TLSRouteStatus) UpdateRouteCondition(condition metav1.Condition) {
+	s.Parents()[0].Conditions = utils.GetNewConditions(s.Parents()[0].Conditions, condition)
+}
```
I cannot come up with a simple way to do that. Seems "generic" in golang don't really fit this part. I create a separate issue. Will follow it up later 
https://github.com/aws/aws-application-networking-k8s/issues/646

zijun726911 commented on pkg/model/core/tlsroute_test.go:
```diff
@@ -0,0 +1,268 @@
+			expectEqual: true,
+		},
+		{
+			description: "Instance populated with the same values are equal",
```
Changed in the new revision

zijun726911 commented on test/pkg/test/httpsapp.go:
```diff
@@ -0,0 +1,81 @@
+	MergeFromService    []*v1.Service
+}
+
+func (env *Framework) NewHttpsApp(options HTTPsAppOptions) (*appsv1.Deployment, *v1.Service) {
```
Sorry, I just found `NewHttpApp` and `NewHttpsApp`
 use 2 different images:

- "public.ecr.aws/x2j8p8w7/https-server:latest" (in which the  https ngix has a cert and listen to 443)

- "public.ecr.aws/x2j8p8w7/http-server:latest" (in which the  http ngix listen to 8090)

Change the `NewHttpApp()` function to make it work for both  "public.ecr.aws/x2j8p8w7/https-server:latest" or "public.ecr.aws/x2j8p8w7/http-server:latest" will cause a large code diff as the `NewHttpApp(test.HTTPAppOptions{` is used in many test cases.

<img width="1276" alt="image" src="https://github.com/aws/aws-application-networking-k8s/assets/32318664/dfdc6cda-d87f-475d-ab95-14380cec61cd">


  I would still use the separate function ` NewHttpsApp(options HTTPsAppOptions)()` and do the refactoring afterward.

erikfuller replied:
ok - thanks for having a look.

Mengli-Shu reviewed (approved):
only reviewed doc changes for lattice tls passthrough behaviour

erikfuller commented on pkg/deploy/lattice/listener_manager.go:
```diff
@@ -45,45 +47,56 @@ func (d *defaultListenerManager) Upsert(
+		Id:          aws.StringValue(latticeListenerSummary.Id),
+		ServiceId:   latticeSvcId,
+	}
+	if modelListener.Spec.Protocol != vpclattice.ListenerProtocolTlsPassthrough {
```
With the refactoring proposed, you should be able to simply compare the default action on the model vs actual and we can drop any special handling for different protocol types here.

zijun726911 replied:
My intention here is to save one api call for `vpclattice.GetListener()` for non-TlsPassthrough listener, not necessary?

erikfuller replied:
OK. I think this is fine as-is.

erikfuller commented on pkg/deploy/lattice/listener_synthesizer.go:
```diff
@@ -46,7 +50,19 @@ func (l *listenerSynthesizer) Synthesize(ctx context.Context) error {
+			}
+			// Fill the default action target groups for TLS_PASSTHROUGH listener, since TLS_PASSTHROUGH listener only has the defaultAction and no extra listener rules
+			defaultActionTgs = stackRules[0].Spec.Action.TargetGroups
+		}
```
Added some more comments on the topic but we can resolve this thread.

erikfuller commented on pkg/gateway/model_build_rule.go:
```diff
@@ -64,12 +68,19 @@ func (t *latticeServiceModelBuildTask) buildRules(ctx context.Context, stackList
 			if _, ok := rule.(*core.GRPCRouteRule); ok {
 				ruleSpec.Method = string(gwv1.HTTPMethodPost)
 			}
+
+			// For TLSRoute, although in here we create a lattice listener rule with PathMatchValue = "/", but it will not be used in the rule_synthesizer,
+			// Since vpc lattice doesn't support any rule matching for TLS_PASSTHROUGH listener,
+			// the rule_synthesizer will skip updating the rule while the listener_synthesizer will fill the listener with the default action route to the referenced  TCP target groups
+			// https://docs.aws.amazon.com/vpc-lattice/latest/ug/tls-listeners.html
```
The reason I'd like to change is that right now we have a few comment lines explaining why the code should drop through to the `else` for TLS passthrough. A multi-line explanation in comments can be a warning sign we don't have the code structured correctly. I'd like us to rather take an approach that is explicit, intuitive to follow, and needs no justifying or explanatory comments. I think what I proposed (or something like it) would do the trick.

Ideally we would redefine the interfaces so there were separate L4 and L7 interfaces, and there would be no `Matches()` at all on TCPRoute. This would be a larger change and is probably OK to skip for now, but we should not be relying on the empty list behavior here.

erikfuller reviewed (changes requested):
Thanks again @zijun726911, for all your work on this. I just have a couple remaining requests to work through then I think we'll be all set.

zijun726911 commented on pkg/deploy/lattice/listener_synthesizer.go:
```diff
@@ -80,6 +89,45 @@ func (l *listenerSynthesizer) Synthesize(ctx context.Context) error {
 	return nil
 }
 
+func (l *listenerSynthesizer) getLatticeListenerDefaultAction(ctx context.Context, modelListenerProtocol string) (
```
Thanks for you comment! I did some changes in the new revision and that did make the code cleaner: 

1. Added the field `DefaultAction` in the `modelListener.ListenerSpec` https://github.com/aws/aws-application-networking-k8s/blob/cd6072c039fed864ec7042ee6e38ab348f1f528d/pkg/model/lattice/listener.go#L13-L32

2. Completely skip to create stackRule for TLS_PASSTHROUGH gwListener. i.e,  completely don't call  `atticeServiceModelBuildTask.buildRules()` for  TLS_PASSTHROUGH gwListener, therefore removing the "long comments" in the `buildRules()`  https://github.com/aws/aws-application-networking-k8s/blob/cd6072c039fed864ec7042ee6e38ab348f1f528d/pkg/gateway/model_build_lattice_service.go#L89-L92

3. Added the ` (t *latticeServiceModelBuildTask) getListenerDefaultAction` in the `model_build_listener` https://github.com/aws/aws-application-networking-k8s/blob/cd6072c039fed864ec7042ee6e38ab348f1f528d/pkg/gateway/model_build_listener.go#L128-L151 (Added unit test to cover build TLS_PASSTHROUGH listener)

4. Changed the logic of `(l *listenerSynthesizer) getLatticeListenerDefaultAction(` https://github.com/aws/aws-application-networking-k8s/blob/cd6072c039fed864ec7042ee6e38ab348f1f528d/pkg/deploy/lattice/listener_synthesizer.go#L92-L127 , removed the seemly confusing part `l.tgManager.ResolveRuleTgIds(ctx, stackRules[0], l.stack)`, change to ` l.tgManager.ResolveRuleTgIds(ctx, stackListener.Spec.DefaultAction.Forward, l.stack);()` instead.

5. Still keep the defaultAction argument for the `listenerManager.Upsert()` , since the in the listenerSynthesizer it need to convert the `modelListenerDefaultAction` into the `vpcLatticeListenerDefaultAction`, and pass the `vpcLatticeListenerDefaultAction` to `listenerManager.Upsert()`  to determine (`needToUpdateDefaultAction()`) whether the listener need to be updated.



Ran the full test suite for the new revision(skip the RAM test):
```
• [257.448 seconds]
------------------------------
[SynchronizedAfterSuite]
/Volumes/workspace/aws-application-networking-k8s/test/suites/integration/suite_test.go:72
{“level”:“info”,“ts”:“2024-06-05T23:08:51.716-0700",“caller”:“test/framework.go:264",“msg”:“Deleting objects: *v1.Gateway/test-gateway, *v1.Pod/grpc-runner”}
{“level”:“info”,“ts”:“2024-06-05T23:08:51.758-0700",“caller”:“test/framework.go:282",“msg”:“Waiting for NotFound, objects: *v1.Gateway/test-gateway, *v1.Pod/grpc-runner”}
[SynchronizedAfterSuite] PASSED [31.193 seconds]
------------------------------
Ran 66 of 69 Specs in 2412.270 seconds
SUCCESS! -- 66 Passed | 0 Failed | 0 Pending | 3 Skipped
--- PASS: TestIntegration (2412.27s)
PASS
```

zijun726911 commented on pkg/controllers/route_controller.go:
```diff
@@ -51,12 +53,12 @@ import (
 var routeTypeToFinalizer = map[core.RouteType]string{
 	core.HttpRouteType: "httproute.k8s.aws/resources",
 	core.GrpcRouteType: "grpcroute.k8s.aws/resources",
+	core.TlsRouteType:  "tlsroute.k8s.aws/resource",
```
Changed in the new revisions

zijun726911 commented:
> 
> The reason I'd like to change is that right now we have a few comment lines explaining why the code should drop through to the else for TLS passthrough. A multi-line explanation in comments can be a warning sign we don't have the code structured correctly. I'd like us to rather take an approach that is explicit, intuitive to follow, and needs no justifying or explanatory comments. I think what I proposed (or something like it) would do the trick.
> 
> Ideally we would redefine the interfaces so there were separate L4 and L7 interfaces, and there would be no Matches() at all on TCPRoute. This would be a larger change and is probably OK to skip for now, but we should not be relying on the empty list behavior here.


Changed in the new revision. please check: https://github.com/aws/aws-application-networking-k8s/pull/643#discussion_r1629937229


> Should we just drop this Expect() then since we don't really care?

Removed assertion for the tg port in the new revision.

erikfuller commented on pkg/deploy/lattice/listener_synthesizer.go:
```diff
@@ -80,6 +89,45 @@ func (l *listenerSynthesizer) Synthesize(ctx context.Context) error {
 	return nil
 }
 
+func (l *listenerSynthesizer) getLatticeListenerDefaultAction(ctx context.Context, stackListener *model.Listener) (
```
The conversion logic from model -> Lattice type typically lives in each component's `manager`. Here, it should only resolve the target group Ids.

Please move `getDefaultAction` logic inside `listener_manager`. The logic should assume the target group Ids have already been resolved. However, it needs to handling the case where the target group ID is set to `model.InvalidBackendRefTgId`. This can happen when the `backendRef` points to something that doesn't exist. Note we should have a test specifically for `InvalidBackendRefTgId` handling, which I don't think we do yet since I don't see it referenced here.

zijun726911 replied:
Working on in. Will do the exactly you describe here in the new revision. Will mimic the `r *defaultRuleManager) buildLatticeRule()` to get the listener defaultAction

 https://github.com/aws/aws-application-networking-k8s/blob/3e6585687487fe3339063b610527403b2dab183a/pkg/deploy/lattice/rule_manager.go#L92-L143

zijun726911 replied:
Sent the new revision can you take a look? Did the changes exactly as you describe above.  Thank you! https://github.com/aws/aws-application-networking-k8s/pull/643/commits/b4dfd4dedcbe6cfce9d935f3477dcf5e2e29b203

> Note we should have a test specifically for InvalidBackendRefTgId handling
The test cases: `Test_ListenerManager_getLatticeListenerDefaultAction_TLS_PASSTHROUGH_Listener` covered that.


Also ran the full e2e test suite against new revision
```
[SynchronizedAfterSuite]
/Volumes/workspace/aws-application-networking-k8s/test/suites/integration/suite_test.go:72
{“level”:“info”,“ts”:“2024-06-07T01:26:52.316-0700",“caller”:“test/framework.go:264",“msg”:“Deleting objects: *v1.Gateway/test-gateway, *v1.Pod/grpc-runner”}
{“level”:“info”,“ts”:“2024-06-07T01:26:52.357-0700",“caller”:“test/framework.go:282",“msg”:“Waiting for NotFound, objects: *v1.Gateway/test-gateway, *v1.Pod/grpc-runner”}
[SynchronizedAfterSuite] PASSED [31.041 seconds]
------------------------------
Ran 66 of 69 Specs in 2407.814 seconds
SUCCESS! -- 66 Passed | 0 Failed | 0 Pending | 3 Skipped
--- PASS: TestIntegration (2407.82s)
PASS
ok 	[github.com/aws/aws-application-networking-k8s/test/suites/integration](http://github.com/aws/aws-application-networking-k8s/test/suites/integration)	2408.753s
```

erikfuller commented on pkg/deploy/lattice/listener_synthesizer.go:
```diff
@@ -80,6 +89,45 @@ func (l *listenerSynthesizer) Synthesize(ctx context.Context) error {
+	}
+
+	// If the listener DefaultAction is not fixed response, for example for TLS_PASSTHROUGH listener, it must be a forward action, fill the forward action target group ids for it
+	if err := l.tgManager.ResolveRuleTgIds(ctx, stackListener.Spec.DefaultAction.Forward, l.stack); err != nil {
+		return nil, fmt.Errorf("failed to resolve rule tg ids, err = %v", err)
+	}
```
Bring this block into the `Synthesize()` method.

zijun726911 replied:
Will change in the new revision

erikfuller commented on pkg/deploy/lattice/listener_synthesizer.go:
```diff
@@ -80,6 +89,45 @@ func (l *listenerSynthesizer) Synthesize(ctx context.Context) error {
+func (l *listenerSynthesizer) getLatticeListenerDefaultAction(ctx context.Context, stackListener *model.Listener) (
+	*vpclattice.RuleAction, error,
+) {
+	if stackListener.Spec.DefaultAction == nil ||
+		(stackListener.Spec.DefaultAction.FixedResponseStatusCode == nil && stackListener.Spec.DefaultAction.Forward == nil) {
+		return nil, fmt.Errorf("invalid listener default action, must be either fixed response or forward")
+	}
```
We should rather validate this when we add the listener to the stack in `listener.go#NewListener()`.

zijun726911 replied:
Will mimic the `func (t *TargetGroupSpec) Validate()`, and put that validation in the new function `func (spec *ListenerSpec) Validate()`

erikfuller reviewed (changes requested):
Just a couple small changes and we should be all set.

erikfuller reviewed (changes requested):
Just a couple very small changes.

zijun726911 commented on pkg/deploy/lattice/listener_manager.go:
```diff
@@ -98,16 +113,81 @@ func (d *defaultListenerManager) Upsert(
+		}
+	}
+	if !hasValidTargetGroup {
+		return &vpclattice.RuleAction{
+			FixedResponse: &vpclattice.FixedResponseAction{
+				StatusCode: aws.Int64(model.DefaultActionFixedResponseStatusCode),
+			},
+		}
```
Ok, I think you mean the listener protocol "TLS_PASSTHROUGH" here, make sense to add that.

and "TCPRoute" ---> "TLSRoute"

![image](https://github.com/aws/aws-application-networking-k8s/assets/32318664/4735c532-55b2-4abc-962e-e2b6cd4514ae)

erikfuller replied:
That's the one.

zijun726911 commented on pkg/deploy/lattice/listener_synthesizer.go:
```diff
@@ -46,6 +50,13 @@ func (l *listenerSynthesizer) Synthesize(ctx context.Context) error {
 			return err
 		}
 
+		if listener.Spec.Protocol == vpclattice.ListenerProtocolTlsPassthrough {
```
make sense, will change

erikfuller commented on pkg/deploy/lattice/listener_manager.go:
```diff
@@ -98,16 +116,81 @@ func (d *defaultListenerManager) Upsert(
+		if stackListener.Spec.Protocol == vpclattice.ListenerProtocolTlsPassthrough {
+			return nil, fmt.Errorf("TLSRoute %s/%s must have at least one valid backendRef target group", stackListener.Spec.K8SRouteNamespace, stackListener.Spec.K8SRouteName)
+		} else {
+			return nil, fmt.Errorf("unreachable code, since the defaultAction for non-TLS_PASSTHROUGH listener is always the FixedResponse 404")
```
Famous last words 😉

erikfuller reviewed (approved):
Thanks so much for all your hard work on this!
